src/Microsoft.ApplicationInsights/Extensibility/Implementation/Endpoints/EndpointProvider.cs

```diff
@@ -0,0 +1,180 @@
+                throw; // Re-throw original exception
+            }
+        }
+
+        /// <summary>
+        /// Will evaluate connection string and return the requested instrumentation key.
+        /// </summary>
+        /// <returns>Returns the instrumentation key from the connection string.</returns>
+        public string GetInstrumentationKey()
+        {
+            if (this.connectionStringParsed.TryGetValue("InstrumentationKey", out string value))
+            {
+                return value;
+            }
+            else
+            {
+                throw new ConnectionStringMissingInstrumentationKeyException();
+            }
+        }
+
+        /// <summary>
+        /// Parse a string connection string and return a Dictionary.
+        /// </summary>
+        /// <remarks>Example: "key1=value1;key2=value2;key3=value3".</remarks>
+        /// <returns>A dictionary parsed from the input connection string.</returns>
+        internal static Dictionary<string, string> ParseConnectionString(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return new Dictionary<string, string>(0);
+            }
+
+            try
+            {
+                return value
+                    .Split(SplitSemicolon, StringSplitOptions.RemoveEmptyEntries)
+                    .Select(part => part.Split('='))
+                    .ToDictionary(split => split[0], split => split[1], StringComparer.OrdinalIgnoreCase);
+            }
+            catch (ArgumentException ex) when (ex.Message.StartsWith("An item with the same key has already been added.", StringComparison.Ordinal))
+            {
+                // TODO: LOG TO ETW ERROR: duplicate keys
+                throw new ConnectionStringDuplicateKeyException("The Connection String has duplicate keys.", ex);
+            }
+            catch (IndexOutOfRangeException ex) when (ex.Message.StartsWith("Index was outside the bounds of the array.", StringComparison.Ordinal))
+            {
+                // TODO: LOG TO ETW ERROR: connection string invalid format
+                throw new ConnectionStringInvalidDelimiterException("The Connection String has invalid formatting and cannot be parsed. Expected: 'key1=value1;key2=value2;key3=value3'", ex);
+            }
+        }
+
+        /// <summary>
+        /// Construct a Uri from the possible parts.
+        /// Will also attempt to sanitize user input.
+        /// </summary>
+        /// <remarks>
+        /// Format: "location.prefix.suffix".
+        /// Example: "https:// westus2.dc.applicationinsights.azure.cn/".
+        /// </remarks>
+        /// <returns>Returns a <see cref="Uri"/> built from the inputs.</returns>
+        internal static Uri BuildUri(string prefix, string suffix, string location = null)
+        {
+            // Location and Host are user input fields and need to be checked for extra periods.
```

Comment:
It would also be bad to have a slash in the prefix or the location (and probably also the suffix)

Reply:
If there are existing functions/methods to do the validation, we should use them.
If there is no existing function, we should follow the RFC and perhaps use regular expression to match the ABNF.

Reply:
The `Uri` class will do validation which is why I want to use it here.
If there's something easy to fix like a trailing slash or period I would like to fix it on the user's behalf.
But there's an argument to be made for failing fast and explicitly when there's bad input.